modules/couchbase-cluster/variables.tf

```diff
@@ -42,72 +49,84 @@ variable "user_data" {
-  type        = "list"
+  type        = list(string)
   default     = []
 }
 
 variable "allowed_ssh_security_group_ids" {
   description = "A list of security group IDs from which the EC2 Instances will allow SSH connections"
-  type        = "list"
+  type        = list(string)
   default     = []
 }
 
 variable "termination_policies" {
   description = "A list of policies to decide how the instances in the auto scale group should be terminated. The allowed values are OldestInstance, NewestInstance, OldestLaunchConfiguration, ClosestToNextInstanceHour, Default."
+  type        = string
   default     = "Default"
 }
 
 variable "associate_public_ip_address" {
   description = "If set to true, associate a public IP address with each EC2 Instance in the cluster."
+  type        = bool
   default     = false
 }
 
 variable "spot_price" {
   description = "The maximum hourly price to pay for EC2 Spot Instances."
+  type        = string
   default     = ""
 }
 
 variable "tenancy" {
   description = "The tenancy of the instance. Must be one of: empty string, default, or dedicated. For EC2 Spot Instances only empty string or dedicated can be used."
-  default     = ""
+  type        = string
+  default     = null
 }
 
 variable "root_volume_ebs_optimized" {
   description = "If true, the launched EC2 instance will be EBS-optimized."
+  type        = bool
   default     = false
 }
 
 variable "root_volume_type" {
   description = "The type of volume. Must be one of: standard, gp2, or io1."
+  type        = string
   default     = "gp2"
 }
 
 variable "root_volume_size" {
   description = "The size, in GB, of the root EBS volume."
+  type        = number
   default     = 50
 }
 
 variable "root_volume_delete_on_termination" {
   description = "Whether the volume should be destroyed on instance termination."
+  type        = bool
   default     = true
 }
 
 variable "root_volume_iops" {
   description = "The amount of provisioned IOPS for the root volume. Only used if volume type is io1."
+  type        = number
   default     = 0
 }
 
 variable "ebs_block_devices" {
   description = "A list of EBS volumes to attach to each EC2 Instance. Each item in the list should be an object with the keys 'device_name', 'volume_type', 'volume_size', 'iops', 'delete_on_termination', and 'encrypted', as defined here: https://www.terraform.io/docs/providers/aws/r/launch_configuration.html#block-devices. We recommend using one EBS Volume for the Couchbase data dir and another one for the index dir."
-  type        = "list"
-  default     = []
-
+  # We can't narrow the inner type down more than "any" because if we use object, then all the fields will be required
+  # (whereas some, such as encrypted, should be optional), and if we use map, all the values must be of the same type,
+  # whereas some here are strings, some are bools, and some are ints.
+  type    = list(any)
```

Comment:
This will not work in all cases. In particular, it will complain if you have:

```
default = [
  {
    device_name = "/dev/xvdh"
    volume_type = "gp2"
  },
  {
    device_name = "/dev/xvdh"
    volume_type = "gp2"
    encrypted = true
  },
]
```

This fails because the second `map` is not the same type as the first `map`. You need to actually use `any`.

Reply:
Oh, good point! https://github.com/gruntwork-io/terraform-aws-couchbase/pull/49/commits/8c9412aa93eda61ab056fb974cd6894d1e232cbf